Update JS debugging learn article (#38840)
mdn/content: files/en-us/learn_web_development/core/scripting/debugging_javascript/index.md

        <ul>
          <li>Using browser developer tools to inspect the JavaScript running on your page and see what errors it is generating.</li>
          <li>Using <code>console.log()</code> and <code>console.error()</code> for debugging.</li>

>           <li>Advanced JavaScript debugging with devtools JavaScript debuggers.</li>

Reviewer: ```suggestion
          <li>Advanced JavaScript debugging with browser devtools.</li>
```

Reviewer: I think it is self evident why, but can discuss if needed ^^^

The `console.log()` output shows that the return value of `fetch()` is not the JSON data, it's a {{jsxref("Promise")}}. The `fetch()` function is asynchronous: it returns a `Promise` that is fulfilled only when the actual response has been received from the network. Before we can use the response, we have to wait for the `Promise` to be fulfilled.

### `console.error()` and call stacks

As a brief digression, lets try using a different console method to report the error — {{domxref("console.error()")}}. In your code, replace

```js
console.log(`Response value: ${response}`);
```

with

```js
console.error(`Response value: ${response}`);
```

> Save and refresh the browser, and you'll now see the message reported as an error, with the same color and icon as the uncaught error below it. In addition, there will now be a expand/collapse arrow next to the message. If you press this, you'll see a single line telling you the line in the JavaScript file the error originated from. In fact, the uncaught error line _also_ has this, but it has two lines:

Reviewer 2: ```suggestion
Save your code and refresh the browser and you'll now see the message reported as an error, with the same color and icon as the uncaught error below it. In addition, there will now be a expand/collapse arrow next to the message. If you press this, you'll see a single line telling you the line in the JavaScript file the error originated from. In fact, the uncaught error line _also_ has this, but it has two lines:
```

```plain
showHeroes http://localhost:7800/js-debug-test/index.js:25
<anonymous> http://localhost:7800/js-debug-test/index.js:10
```

> This means that the error is coming from the `showHeroes()` function, line 25, as we noted earlier. If you look at your code, you'll see that the anonymous call on line 10 is the line that is calling `showHeroes()`. This is called a **call stack**, and it can be really useful when trying to track down the source of an error involving lots of different locations in your code.

Reviewer: Not sure "this" is clear

```suggestion
This means that the error is coming from the `showHeroes()` function, line 25, as we noted earlier. If you look at your code, you'll see that the anonymous call on line 10 is the line that is calling `showHeroes()`. These lines are referred to as a **call stack**, and can be really useful when trying to track down the source of an error involving lots of different locations in your code.
```

> OK, so `console.error()` didn't help us _that_ much here, but it is useful to be able to generate a custom call stack if one is not available.

Reviewer: I'm not really sure what this means - i.e. what's a custom call stack? Why wouldn't we get one etc. Doesn't fix the problem, but something like this might be OK.

```suggestion
The `console.error()` call isn't all that useful in this case, but it can be useful to be able to generate a custom call stack if one is not available.
```

Author: I think really this is just a call stack, and using the word "custom" has confused things rather. I've update this line to:

> The `console.error()` call isn't all that useful in this case, but it can be useful for generating a call stack if one is not already available.

### Fixing the error

> Anyway, let's get back to trying to fix our error. We can access the response from the fulfilled `Promise` by chaining the {{jsxref("Promise.prototype.then()", "then()")}} method onto the end of the `fetch()` call. We can then pass the resulting response value into the functions that accept it, like this:

Reviewer: Should we be updating to async/await?

Author: Probably, but I don't want to do that in this PR.

});
```

Save and refresh, and see if your code is working. Spoiler alert — the above change has not fixed the problem. Unfortunately, we **still have the same error**!

> [!NOTE]

> > To summarize, anytime something is not working and a value does not appear to be what it is meant to be at some point in your code, you can use `console.log()`, `console.error()`, or another similar function to print out the value and see what is happening.

Reviewer: Is anytime a word?

Reviewer: ```suggestion
> To summarize, any time something is not working and a value does not appear to be what it is meant to be at some point in your code, you can use `console.log()`, `console.error()`, or another similar function to print out the value and see what is happening.
```

The main feature of such tools is the ability to add breakpoints to code — these are points where the execution of the code stops, and at that point you can examine the environment in its current state and see what is going on.

> Let's get to work. The error is being thrown at the same line as before — `for (const hero of heroes) {` — line 26 in the screenshot below. Click on this line in the center panel to add a breakpoint to it (you'll see a blue arrow appear over the top of it). Now refresh the page (<kbd>Cmd</kbd>/<kbd>Ctrl</kbd> + <kbd>R</kbd>) — the browser will pause execution of the code on that line. At this point, the right-hand side will update to show the following:

Reviewer: FWIW I don't like the conversational style in learn "Let's get to work. ". Shan't!

## Handling JavaScript errors in your code

> HTML and CSS are permissive — errors and unrecognized features can often be handled due to the nature of the languages. For example, CSS will ignore unrecognized properties, and the rest of the code will often just work. JavaScript is not as permissive as HTML and CSS however — if the JavaScript engine encounters mistakes or unrecognized syntax, it will often throw errors.

Reviewer 2: Do you want to mention strict mode here?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode

Author: I don't think mentioning strict mode would be useful here; it would probably confuse beginners more than it helps.

   ```js
   try {
     console.log(inchesToMeters(height));
   } catch (error) {
     console.error(error);
     console.log("Insert code to handle the error");
   }
   ```

2. Save and refresh, and you should now see two things:

   - The error message and call stack as before, but this time, without a label of "uncaught", or "unhandled".
   - The logged message "Insert code to handle the error".

3. Now try updating `num` to a good (number) value, and you'll see the result of the calculation logged, with no error message.

This is significant — any thrown errors are no longer unhandled, so they won't bring the application crashing to a halt. You can run whatever code you like to handle the error. Above we are just logging a message, but for example you could call whatever function was run earlier to ask the user to enter their height, this time asking them to correct the input error. You could even use an `if...else` statement to run different error handling code depending on what type of error is returned.

### Feature detection

> Feature detection is useful when you are planning to use new JavaScript features that might not be supported in all browsers. Test for the feature, and then conditionally run code to provide an acceptable experience both in browsers that do and don't support the feature. As a quick example, the [Geolocation API](/en-US/docs/Web/API/Geolocation_API) (which exposes available location data for the device the web browser is running on) has a main entry point for its use — a `geolocation` property available on the global [Navigator](/en-US/docs/Web/API/Navigator) object. Therefore, you can detect whether the browser supports geolocation or not by using a similar `if()` structure to what we saw earlier:

Reviewer 2: I recently added some examples here if there's anything interesting to bring in to this page: https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Browser_detection_using_the_user_agent#alternatives_to_ua_sniffing

Author: Good call; I added a line below the example linking to this section.

Reviewer 2: Nice, thanks 👍🏻
